File: py/src/apconf/process_yaml.py
"""Read config data from yaml files in given dirs. Support params."""
from collections import defaultdict
from pathlib import Path
import yaml
from jinja2 import Template

def process_yaml_dirs(dirs, template_params):
    """Process dirs."""
    final_dict = defaultdict(dict, {})

    def merge_dicts(d1, d2):
        for key in d2:
            if key in d1:
                if isinstance(d1[key], dict) and isinstance(d2[key], dict):
                    merge_dicts(d1[key], d2[key])
                elif isinstance(d1[key], list) and isinstance(d2[key], list):
                    d1[key] += d2[key]
                elif d1[key] != d2[key]:
                    # pylint:disable=broad-exception-raised
                    raise Exception(
                        f"Conflict in merging dictionaries at key: {key}")
            else:
                d1[key] = d2[key]

    def process_yaml_content(content):
        loaded_content = yaml.safe_load_all(content)
        for doc in loaded_content:
            if 'kind' not in doc or 'metadata' not in doc or 'name' not in doc['metadata']:
                continue
            kind = doc['kind']
            name = doc['metadata']['name']
            wrapped = {kind: {name: doc}}
            merge_dicts(final_dict, wrapped)

    for dir_path in dirs:
        for file_path in Path(dir_path).rglob('*.yaml'):
            if file_path.name.startswith ('.'):
                continue
            with open(file_path, 'r', encoding='utf-8') as file:
                process_yaml_content(
                    Template(file.read()).render(template_params))
    return dict(final_dict)

File: go/apconf/process_yaml.go
package apconf

import (
	"bytes"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"text/template"
	"unicode"

	"gopkg.in/yaml.v3"
)

func processYamlDirs(dirs []string, templateParams map[string]any) map[string]any {
	finalDict := make(map[string]any)

	for _, dirPath := range dirs {
		entries, err := os.ReadDir(dirPath)
		if err != nil {
			panic(err) // Handle error as needed
		}

		for _, entry := range entries {
			if entry.IsDir() || strings.HasPrefix(entry.Name(), ".") || filepath.Ext(entry.Name()) != ".yaml" {
				continue
			}

			content, err := os.ReadFile(filepath.Join(dirPath, entry.Name()))
			if err != nil {
				panic(err) // Handle error as needed
			}
			processedContent := renderTemplate(content, templateParams)
			processYamlContent(processedContent, finalDict)
		}
	}
	return finalDict
}

func processYamlContent(content []byte, finalDict map[string]any) {
	// First try to unmarshal into a single document map
	var singleDoc map[string]any
	if err := yaml.Unmarshal(content, &singleDoc); err == nil {
		mergeYamlDocument(singleDoc, finalDict)
		return
	}

	// If unmarshalling into a single map failed, try a slice of maps
	var multiDocs []map[string]any
	if err := yaml.Unmarshal(content, &multiDocs); err == nil {
		for _, doc := range multiDocs {
			mergeYamlDocument(doc, finalDict)
		}
	} else {
		// Handle the error properly (log, panic, or return an error)
		panic(err)
	}
}

func mergeYamlDocument(doc map[string]any, finalDict map[string]any) {
	if kind, ok := doc["kind"].(string); ok {
		if metadata, ok := doc["metadata"].(map[string]any); ok {
			if name, ok := metadata["name"].(string); ok {
				if finalDict[kind] == nil {
					finalDict[kind] = make(map[string]any)
				}
				finalDict[kind].(map[string]any)[name] = doc
			}
		}
	}
}

func preprocessTemplateForGo(content []byte) []byte {
	// This regular expression matches '{{ var_name }}' and captures 'var_name'
	re := regexp.MustCompile(`{{\s*(\w+)\s*}}`)

	// Replace '{{ var_name }}' with '{{ .VarName }}'
	processedContent := re.ReplaceAllFunc(content, func(match []byte) []byte {
		// Extract the variable name
		varName := re.FindSubmatch(match)[1]
		// Convert to Go style '{{ .VarName }}'
		pascalVarName := toPascalCase(string(varName))
		if !strings.HasPrefix(pascalVarName, ".") {
			pascalVarName = "." + pascalVarName
		}
		return []byte("{{ " + pascalVarName + " }}")
	})

	return processedContent
}

// Helper function to convert a string from snake_case to PascalCase
func toPascalCase(s string) string {
	parts := strings.Split(s, "_")
	for i, part := range parts {
		if len(part) > 0 {
			// Convert the first character to uppercase
			runes := []rune(part)
			runes[0] = unicode.ToUpper(runes[0])
			// Convert the rest of the characters to lowercase (optional)
			parts[i] = string(runes)
		}
	}
	// Join all parts together to form PascalCase
	return strings.Join(parts, "")
}

func renderTemplate(content []byte, templateParams map[string]any) []byte {
	content = preprocessTemplateForGo(content)

	// Create a new template and parse the content into it.
	tmpl, err := template.New("configTemplate").Parse(string(content))
	if err != nil {
		panic(err)
	}

	// Use a buffer to capture the output of the template execution.
	var renderedContent bytes.Buffer
	err = tmpl.Execute(&renderedContent, templateParams)
	if err != nil {
		panic(err)
	}

	return renderedContent.Bytes()
}
